lib/elixir/pages/Library Guidelines.md

```diff
@@ -0,0 +1,214 @@
+# Library Guidelines
+
+This document outlines general guidelines, anti-patterns, and rules for those writing and publishing Elixir libraries.
+
+## Getting started
+
+You can create a new Elixir library by running the `mix new` command:
+
+    $ mix new my_library
+
+The project name is given in the `snake_case` convention where all letters are lowercase and words are separate with underscores. This is the same convention used by variables, function names and atoms in Elixir. See the [Naming Conventions](naming-conventions.html) document for more information.
+
+Every project has a `mix.exs` file, with instructions on how to build, compile, run tests, and so on. Libraries commonly have a `lib` directory, which includes Elixir source code, and a `test` directory. A `src` directory may also exist for Erlang sources.
+
+For more information on running your project, see the official [Mix & OTP guide](https://elixir-lang.org/getting-started/mix-otp/introduction-to-mix.html) or [Mix documentation](https://hexdocs.pm/mix/Mix.html).
+
+### Applications with supervision tree
+
+The `mix new` command also allows the `--sup` flag to scaffold an application with a supervision tree out of the box. We talk about supervision trees later on when discussing one of the common anti-patterns when writing libraries.
+
+## Publishing
+
+Writing code is only the first of many steps to publish a package. We strongly recommend developers to:
+
+  * Choose a versioning schema. Elixir requires versions to be in the format `MAJOR.MINOR.PATCH` but the meaning of those numbers is up to you. Most projects choose [Semantic Versioning](https://semver.org/).
+
+  * Write tests. Elixir ships with a test-framework named [ExUnit](https://hexdocs.pm/ex_unit/ExUnit.html) and the project generated by `mix new` includes sample tests and doctests.
+
+  * Write documentation. The Elixir community is proud in treating documentation as a first-class citizen and making documentation easily accessible. Libraries contribute to the status quo by providing complete API documentation with examples for their modules, types and functions. Projects like [ExDoc](https://github.com/elixir-lang/ex_doc) can be used to generate HTML and EPUB documents from the documentation. ExDoc also supports writing "extra pages", like this one that you are reading, that augment the documetnation with tutorials, guides, and references. See the [Writing Documentation](writing-documentation.html) page for more information.
```

Comment:
I'm not sure what core's opinion of typespecs is but I find they are very beneficial to the documentation and for users of typespecs it's pretty nice if at least your major public functions have type specs.

Reply:
Typespecs are a great but without the accompanying tooling we can't even make sure they are correct. I need to catch up on the tooling then we can push them as best practices.